### PR TITLE
feat: split rdbms into individual blocks and two challengeTypes

### PR DESCRIPTION
**Pull request by moT01**
This does three things:

1. Splits the RDBMS into two challengeTypes
-One for the practice projects, one for cert projects

2. Splits the RDBMS into individual blocks
-Created files, intros, and text for all of the blocks across all i18n’s. This is where most of the changes are from.

3. Makes it so a URL is required to complete the cert projects
-This adds a `partiallyCompetedChallenges` array to the db. After a user completes a cert project in the VM, the info for it will be added to that array.
-Once they complete the project in the VM, they will be allowed to submit a URL to a repo for it. Submitting a URL will delete the challenge from `partiallyCompletedChallenges` and add it to `completedChallenges`

Completing projects in the VM won’t save to your local database - if you want to do that, make a post request to `/coderoad-challenge-complete` with this info:  
```
/* Example request coming from CodeRoad:
   * req.body: { tutorialId: 'freeCodeCamp/learn-bash-by-building-a-boilerplate:v1.0.0' }
   * req.headers: { coderoad-user-token: '8kFIlZiwMioY6hqqt...' }
   */
```
 Using the repo name in the `tutorialId` and the webhook token you create locally as the `coderoad-user-token`

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

File: client/src/templates/Challenges/codeally/show.tsx [hunk 1]
```diff
@@ -1,53 +1,129 @@
...
   static displayName: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private _container: any;
+  constructor(props: ShowCodeAllyProps) {
+    super(props);
+    this.state = {
+      showIframe: false
+    };
+    this.showIframe = this.showIframe.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/codeally/show.tsx:120]:
```suggestion
```
We can just use arrow functions below and not worry about binds.

File: client/src/templates/Challenges/codeally/show.tsx [hunk 2]
```diff
@@ -56,27 +132,96 @@ class ShowCodeAlly extends Component<ShowCodeAllyProps> {
...
+    this.setState({
+      showIframe: true
+    });
+  }
+
+  handleSubmit({
+    showCompletionModal
+  }: {
+    showCompletionModal: boolean;
+  }): void {
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/codeally/show.tsx:149]:
```suggestion
  handleSubmit = ({
    showCompletionModal
  }: {
    showCompletionModal: boolean;
  }) => {
```

File: client/src/templates/Challenges/codeally/show.tsx [hunk 3]
```diff
@@ -87,25 +232,162 @@ class ShowCodeAlly extends Component<ShowCodeAllyProps> {
...
+                )}
+                <div
+                  className={`ca-btn-padding ${
+                    !isSignedIn || challengeType === 12 ? 'ca-btn-margin' : ''
+                  }`}
+                >
+                  <Button
+                    block={true}
+                    bsStyle='primary'
+                    // eslint-disable-next-line @typescript-eslint/unbound-method
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/codeally/show.tsx:306]:
```suggestion
```

File: client/src/templates/Challenges/codeally/show.tsx [hunk 4]
```diff
@@ -87,25 +232,162 @@ class ShowCodeAlly extends Component<ShowCodeAllyProps> {
...
+                    <div className='ca-description'>
+                      {t('learn.submit-public-url')}
+                    </div>
+                    <Spacer />
+                    <PrismFormatted text={notes} />
+                    <Spacer />
+                    <SolutionForm
+                      challengeType={challengeType}
+                      description={description}
+                      // eslint-disable-next-line @typescript-eslint/unbound-method
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/codeally/show.tsx:338]:
```suggestion
```

File: client/src/templates/Challenges/codeally/show.tsx [hunk 5]
```diff
@@ -87,25 +232,162 @@ class ShowCodeAlly extends Component<ShowCodeAllyProps> {
...
+                      onSubmit={this.handleSubmit}
+                      updateSolutionForm={updateSolutionFormValues}
+                    />
+                  </>
+                )}
+                {/*(!isSignedIn || challengeType === 12) && <Spacer size={2} />*/}
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/codeally/show.tsx:344]:
```suggestion
```
Can this go?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
One minor style issue, the notes look a little different to the rest:

image

File: api-server/src/server/boot/challenge.js [hunk 6]
```diff
@@ -134,6 +137,10 @@ export function buildUserUpdate(
     )
   };
 
+  updateData.$pull = {
+    partiallyCompletedChallenges: { id: challengeId }
+  };
```
[the remark this answers is not in the document]
Comment (moT01) [reply api-server/src/server/boot/challenge.js:142]: Going to mark this resolved, let me know if you think something should be changed here.

File: api-server/src/common/models/user.js [hunk 7]
```diff
@@ -995,6 +995,25 @@ export default function initializeUser(User) {
     });
   };
 
+  User.prototype.getPartiallyCompletedChallenges$ =
+    function getPartiallyCompletedChallenges$() {
+      if (
+        Array.isArray(this.partiallyCompletedChallenges) &&
+        this.partiallyCompletedChallenges.length
+      ) {
```
[the remark this answers is not in the document]
Comment (moT01) [reply api-server/src/common/models/user.js:1003]: Not sure - I suppose if it's not an array, the length check may fail, which might not cause any problems. I just copied what was in the completeChallenges function above this.

[the remark this answers is not in the document]
Comment (moT01) [reply api-server/src/common/models/user.js:1003]: Going to mark this resolved. Let me know if you think something needs to be changed here.

File: api-server/src/common/models/user.js [hunk 8]
```diff
@@ -995,6 +995,25 @@ export default function initializeUser(User) {
...
+        return Observable.of(this.partiallyCompletedChallenges);
+      }
+      const id = this.getId();
+      const filter = {
+        where: { id },
+        fields: { partiallyCompletedChallenges: true }
+      };
+      return this.constructor.findOne$(filter).map(user => {
+        this.partiallyCompletedChallenges = user.partiallyCompletedChallenges;
```
[the remark this answers is not in the document]
Comment (moT01) [reply api-server/src/common/models/user.js:1012]: not sure - again, just copied the completedChallenges function above. It does seem odd though, doesn't it. Maybe something to do with those observables that I never quite figured out how they work.

Going to mark this resolved. Let me know if you think something should be done here.

**Comment by moT01** [comment]
I squashed the commits, rebased, and fixed a few things. This should be good to go again 🤞

File: client/src/templates/Challenges/codeally/show.tsx [hunk 9]
```diff
@@ -87,25 +229,159 @@ class ShowCodeAlly extends Component<ShowCodeAllyProps> {
...
+                  className={`ca-btn-padding ${
+                    !isSignedIn || challengeType === 12 ? 'ca-btn-margin' : ''
+                  }`}
+                >
+                  <Button
+                    block={true}
+                    bsStyle='primary'
+                    onClick={this.showIframe}
+                  >
+                    {challengeType === 13
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/codeally/show.tsx:305]:
```suggestion
                    {challengeType === challengeTypes.codeAllyCert
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1: 

Thanks for bearing with me, @moT01!

File: client/utils/challenge-types.js [hunk 10]
```diff
@@ -11,7 +11,8 @@ const quiz = 8;
 const invalid = 9;
 const pythonProject = 10;
 const video = 11;
-const codeally = 12;
+const codeAllyPractice = 12;
+const codeAllyCert = 13;
```
Comment (ShaunSHamilton) [inline client/utils/challenge-types.js:15]: It seems like we have both `codeAlly` challenge types, as well as `codeRoad` challenge types.

Edit: Never mind, it is just the comments that mention CodeRoad.

**Review by ShaunSHamilton — APPROVED** [review]
LTMG 👍 

I fought everything I had to not suggest changes like `[2, 4, 5, 13].include(challenge.challengeType)`, because that is not Tom's fault... _it is Oliver's_ 😁

**Pull request merged**